src/lib.rs

```diff
@@ -445,6 +446,84 @@ impl<T> Arena<T> {
         }
     }
 
+    /// Get a pair of exclusive references to the elements at index `i1` and `i2` if it is in the
+    /// arena.
+    ///
+    /// If the element at index `i1` or `i2` is not in the arena, then `None` is returned for this
+    /// element.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `i1` and `i2` are pointing to the same item of the arena.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use generational_arena::Arena;
+    ///
+    /// let mut arena = Arena::new();
+    /// let idx1 = arena.insert(0);
+    /// let idx2 = arena.insert(1);
+    ///
+    /// {
+    ///     let (item1, item2) = arena.get2_mut(idx1, idx2);
+    ///
+    ///     *item1.unwrap() = 3;
+    ///     *item2.unwrap() = 4;
+    /// }
+    ///
+    /// assert_eq!(arena[idx1], 3);
+    /// assert_eq!(arena[idx2], 4);
+    /// ```
+    pub fn get2_mut(&mut self, i1: Index, i2: Index) -> (Option<&mut T>, Option<&mut T>) {
+        let len = self.items.len();
+        let ptr = self.items.as_mut_ptr();
+
+        assert!(i1.index != i2.index);
```

Comment:
I don't think this is a valid assertion. We can assert `i1 != i2`, but it should be fine for them to have the same index if their generations are different. The index with the smaller generation should return a `None` (we'll have to handle this case before splitting the array below).

Reply:
Ok, I will change for `split_at_mut`.
About the assertion, its purpose is to not get twice as mutable the element in the vec. In this case it would break the single ownership principle and make the function unsafe.

Reply:
I understand why the *check* needs to be there to avoid mutable aliasing, but it is possible to pass two `Index`es which are distinct but happen to have the same `.index` and we shouldn't panic in this situation. Instead we should figure out which one has the older, invalid generation, and return `None` for that one, and do a normal `get_mut` for the one with the newer generation.

```rust
let i1 = arena.alloc("foo");
arena.remove(i1);
let i2 = arena.alloc("bar"); // will use same index as i1
arena.get2_mut(i1, i2); // this should _not_ panic, but currently will because of the above assertion
```

Come to think of it, we should have the above snippet as a test.

Reply:
I see, I've made the change @fitzgen. I still left the assertion in the case the index and the generation are the same. Tell me what you think :)